geopandas/sindex.py

```diff
@@ -1,25 +1,235 @@
-from geopandas import base
+from warnings import warn
+from collections import namedtuple
 
-if base.HAS_SINDEX:
-    from rtree.index import Index as RTreeIndex
+from pandas import notnull
+import numpy as np
 
+from . import _compat as compat
 
-class SpatialIndex(RTreeIndex):
+
+def has_sindex():
+    """
+    Dynamically checks for ability to generate spatial index.
+    """
+    return get_sindex_class() is not None
+
+
+def get_sindex_class():
     """
-    A simple wrapper around rtree's RTree Index
+    Dynamically chooses a spatial indexing backend.
+    Required to comply with _compat.USE_PYGEOS.
+    The order of preference goes PyGeos > RTree > None.
     """
+    if compat.USE_PYGEOS:
+        return PyGEOSSTRTreeIndex
+    elif compat.HAS_RTREE:
+        return RTreeIndex
+    else:
+        warn("Cannot generate spatial index: Missing package `rtree` or `pygeos`.")
+        return None
+
+
+if compat.HAS_RTREE:
+
+    from rtree.index import Index as ToblerityRTreeIndex  # noqa
+    from rtree.core import RTreeError  # noqa
+    from shapely.geometry import box
+
+    class RTreeIndex(ToblerityRTreeIndex):
+        """
+        A simple wrapper around rtree's RTree Index
+        """
+
+        def __init__(self, geometry):
+            stream = (
```

Comment:
We might want to keep some compatibility here in case `geometry` is already a stream (eg only do this stream creating if `geometry` is a GeoSeries)

Reply:
I am not sure what you mean. I copied this from[ `base.py`](https://github.com/geopandas/geopandas/blob/de58716e79602d63dff8622aa31feadd327a7e31/geopandas/base.py#L105), where it previously lived. I guess in that context it was _always_ a GeoSeries. Are you suggesting we generalize in these wrappers and account for other types of input (ex, array-like of geometries)?